### PR TITLE
[DPC-3683] Initial Client Code for CPI API Gateway

### PR DESCRIPTION
**Pull request by chris-ronning-ny**
## 🎫 Ticket

https://jira.cms.gov/browse/DPC-3683

## 🛠 Changes

This PR establishes a client class that, upon instantiation, authenticates with CMS IDM in the Val environment. As API calls are implemented, this class will continue to manage the access token.

## ℹ️ Context for reviewers

In order to implement an RSpec test for this using VCR, as discussed, the Docker container I'm using needs to be able to connect to the VPN via Zscaler, so that it can make the appropriate API calls. Given that I haven't been able to make this happen, I've opted to exclude unit tests from this PR. Tests will be added in DPC-3684.

## ✅ Acceptance Validation

(How were the changes verified? Did you fully test the acceptance criteria in the ticket? Provide reproducible testing instructions and screenshots if applicable.)

## 🔒 Security Implications

- [ ] This PR adds a new software dependency or dependencies.
- [ ] This PR modifies or invalidates one or more of our security controls.
- [ ] This PR stores or transmits data that was not stored or transmitted before.
- [ ] This PR requires additional review of its security implications for other reasons.

If any security implications apply, add Jason Ashbaugh (GitHub username: StewGoin) as a reviewer and do not merge this PR without his approval.

### CODE REVIEW

**Comment by chris-ronning-ny** [comment]
@jdettmannnava I think we should create a separate ticket for Ansible. It's not really a blocker at this point, so no reason to expand the scope.

File: dpc-portal/docker/entrypoint.sh [hunk 1]
```diff
@@ -10,6 +10,9 @@ fi
 if [ "$1" == "portal" ]; then
   # Start the database service (and make accessible outside the Docker container)
   echo "Starting Rails server..."
+  
+  echo "Migrating the database..."
+  bundle exec rails db:migrate
```
Comment (meliGuzman) [inline dpc-portal/docker/entrypoint.sh:15]: Ruby newbie Q: Why is this needed? Doesn't look like you're touching the db in your code here?

File: dpc-portal/spec/lib/cpi_api_gw_client_spec.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+      client_credentials_strategy_instance = instance_double(OAuth2::Strategy::ClientCredentials)
+      allow(OAuth2::Client).to receive(:new).and_return(oauth_client)
+      allow(oauth_client).to receive(:client_credentials).and_return(client_credentials_strategy_instance)
+      expect(client_credentials_strategy_instance).to receive(:get_token).and_return('test_token')
+
+      client = CPIAPIGatewayClient.new
+      expect(client.token).to eq('test_token')
+    end
+  end
+end
```
Comment (meliGuzman) [inline dpc-portal/spec/lib/cpi_api_gw_client_spec.rb:19]: I'm new to ruby so this file throws me off a bit. Is this a way to set methods/functions for your class?  I saw that  "it" is like a method inside a class like how we have them in java. Is this also sort of testing since you have that expect assertion?

File: dpc-portal/docker/entrypoint.sh (hunk 1 above)
Comment (kyeah) [reply dpc-portal/docker/entrypoint.sh:15]: This is not really related to the PR/ticket but this line was missing from our initial application setup. We need to run database migrations before starting the Rails server, otherwise Rails will complain 😄

File: dpc-portal/spec/lib/cpi_api_gw_client_spec.rb (hunk 2 above)
Comment (meliGuzman) [reply dpc-portal/spec/lib/cpi_api_gw_client_spec.rb:19]: Thank you, Chris!!!

File: dpc-portal/lib/cpi_api_gw_client/cpi_api_gw_client.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+# A client for requests to the CPI API Gateway
+class CPIAPIGatewayClient
+  attr_accessor :token
+
+  def initialize
+    client_id = ENV.fetch('CPI_API_GW_CLIENT_ID', nil)
+    client_secret = ENV.fetch('CPI_API_GW_CLIENT_SECRET', nil)
+    @client = OAuth2::Client.new(client_id, client_secret,
+                                 site: 'https://impl.idp.idm.cms.gov/',
+                                 token_url: '/oauth2/aus2151jb0hszrbLU297/v1/token')
```
Comment (MEspositoE14s) [inline dpc-portal/lib/cpi_api_gw_client/cpi_api_gw_client.rb:14]: My only question is if it will be a problem if `site` ends with a "/" and `token_url` begins with one?  Will that make the final token url `https://impl.idp.idm.cms.gov//oauth2/aus2151jb0hszrbLU297/v1/token`?

**Review by MEspositoE14s — APPROVED** [review]
I don't know how qualified I am to review Rails code, but it seems pretty straightforward and all the tests work, so looks good to me!

File: dpc-portal/lib/cpi_api_gw_client/cpi_api_gw_client.rb (hunk 3 above)
Comment (chris-ronning-ny) [reply dpc-portal/lib/cpi_api_gw_client/cpi_api_gw_client.rb:14]: Nope, this is per spec. See https://gitlab.com/oauth-xx/oauth2#customize-authorize_url-and-token_url
